User management

K-Hir-pk18 opened the pull request:
WHAT
・DEVISE導入
・ユーザー管理機能の実装
・モデルの単体テストコード

WHY
・出品や購入をユーザーと
紐つけるため

K-Hir-pk18 committed:
Basic認証を導入

K-Hir-pk18 committed:
空のcommit

K-Hir-pk18 committed:
bundle install -v2.1.4

K-Hir-pk18 committed:
ユーザー管理機能

WHAT
・DEVISE導入
・ユーザー管理機能の実装
・モデルの単体テストコード作成

WHY
・出品や購入をユーザーと
紐つけるため

haramiyo commented on app/controllers/application_controller.rb:
```diff
@@ -1,10 +1,27 @@
 class ApplicationController < ActionController::Base
 before_action :basic_auth
+before_action :authenticate_user!
```
ApplicationControllerにauthenticate_user!を記述すると、全てのコントローラーのアクションに対してauthenticate_user!が発動してしまいます。
authenticate_user!は各コントローラーでonlyやexpectオプションを使用して、限定的に使用できるようにしたいため、ApplicationControllerでは削除しておきましょう。

haramiyo commented on app/controllers/application_controller.rb:
```diff
@@ -1,10 +1,27 @@
+
+protected
+
+def configure_permitted_parameters
+  devise_parameter_sanitizer.permit(
+    :sign_up, keys: [:nickname, :email, :birth_date, :first_name, :last_name, :first_name_kana, :last_name_kana]
+  )
+end
```
`:email`はsign_upの方で許可できているため不要な記述となります。取り除きましょう。

haramiyo commented on app/controllers/application_controller.rb:
```diff
@@ -1,10 +1,27 @@
+      username == ENV["BASIC_AUTH_USER"] && password == ENV["BASIC_AUTH_PASSWORD"]
+     end
+
+     def configure_permitted_parameters
+      devise_parameter_sanitizer.permit(:sign_up, keys: [:nickname])
+
     end
```
こちあら必要な記述でしょうか？
もし不要であれば削除しておきましょう。

haramiyo commented on app/controllers/products_controller.rb:
```diff
@@ -1,5 +1,47 @@
+  def move_to_index
+    unless user_signed_in?
+      redirect_to action: :index
+    end
```
こちらは一旦コメントアウトしておきましょう！

*理由*
こちらは商品出品や商品一覧に関する実装となります。
現在記述してしまいますと、「商品出品ブランチ」や「商品一覧ブランチ」を作成した際に差分としてプルリクエストに反映されなくなってしまいます。
１ブランチ１機能の実装で進めるようにしましょう！

haramiyo commented on app/models/category.rb:
```diff
@@ -0,0 +1,19 @@
+class category < ActiveHash::Base
+  self.data = [
+    { id: 0, name: '--' },
+    { id: 1, name: 'レディース' },
+    { id: 2, name: 'メンズ' },
+    { id: 3, name: 'ベビー・キッズ' },
+    { id: 4, name: 'インテリア・住まい・小物' },
+    { id: 5, name: '本・音楽・ゲーム' },
+    { id: 6, name: 'おもちゃ・ホビー・グッズ' },
+    { id: 7, name: '家電・スマホ・カメラ' },
+    { id: 8, name: 'スポーツ・レジャー' },
+    { id: 9, name: 'ハンドメイド' },
+    { id: 10, name: 'その他' }
+  ]
+ 
+   include ActiveHash::Associations
+   has_many :products
```
こちらも商品出品機能の実装となりますので、一旦コメントアウトしておきましょう！

haramiyo commented on app/models/condition.rb:
```diff
@@ -0,0 +1,15 @@
+class condition < ActiveHash::Base
+  self.data = [
+    { id: 0, name: '--' },
+    { id: 1, name: '新品、未使用' },
+    { id: 2, name: '未使用に近い' },
+    { id: 3, name: '目立った傷や汚れなし' },
+    { id: 4, name: 'やや傷や汚れあり' },
+    { id: 5, name: '傷や汚れあり' },
+    { id: 6, name: '全体的に状態が悪い' }
+  ]
+ 
+   include ActiveHash::Associations
+   has_many :products
```
こちらも商品出品機能の実装となりますので、一旦コメントアウトしておきましょう！

chobi555 replied:
こちらの記述は本実装とは関係のないものになるので、一度コメントアウトしておきましょう！

haramiyo commented on app/models/prefecture.rb:
```diff
@@ -0,0 +1,25 @@
+class Prefecture < ActiveHash::Base
+  self.data = [
+               {id: 0, name: '--'}, {id: 1, name: '北海道'}, {id: 2, name: '青森県'}, 
+               {id: 3, name: '岩手県'}, {id: 4, name: '宮城県'}, {id: 5, name: '秋田県'}, 
+               {id: 6, name: '山形県'}, {id: 7, name: '福島県'}, {id: 8, name: '茨城県'}, 
+               {id: 9, name: '栃木県'}, {id: 10, name: '群馬県'}, {id: 11, name: '埼玉県'}, 
+               {id: 12, name: '千葉県'}, {id: 13, name: '東京都'}, {id: 14, name: '神奈川県'}, 
+               {id: 15, name: '新潟県'}, {id: 16, name: '富山県'}, {id: 17, name: '石川県'}, 
+               {id: 18, name: '福井県'}, {id: 19, name: '山梨県'}, {id: 20, name: '長野県'}, 
+               {id: 21, name: '岐阜県'}, {id: 22, name: '静岡県'}, {id: 23, name: '愛知県'}, 
+               {id: 24, name: '三重県'}, {id: 25, name: '滋賀県'}, {id: 26, name: '京都府'}, 
+               {id: 27, name: '大阪府'}, {id: 28, name: '兵庫県'}, {id: 29, name: '奈良県'}, 
+               {id: 30, name: '和歌山県'}, {id: 31, name: '鳥取県'}, {id: 32, name: '島根県'}, 
+               {id: 33, name: '岡山県'}, {id: 34, name: '広島県'}, {id: 35, name: '山口県'}, 
+               {id: 36, name: '徳島県'}, {id: 37, name: '香川県'}, {id: 38, name: '愛媛県'}, 
+               {id: 39, name: '高知県'}, {id: 40, name: '福岡県'}, {id: 41, name: '佐賀県'}, 
+               {id: 42, name: '長崎県'}, {id: 43, name: '熊本県'}, {id: 44, name: '大分県'}, 
+               {id: 45, name: '宮崎県'}, {id: 46, name: '鹿児島県'}, {id: 47, name: '沖縄県'}
+              ]
+end
+ 
+   include ActiveHash::Associations
+   has_many :products
```
こちらも商品出品機能の実装となりますので、一旦コメントアウトしておきましょう！

chobi555 replied:
こちらの記述は本実装とは関係のないものになるので、一度コメントアウトしておきましょう！

haramiyo commented on app/models/product.rb:
```diff
@@ -0,0 +1,24 @@
+class Product < ApplicationRecord
+  extend ActiveHash::Associations::ActiveRecordExtensions
+  belongs_to :category
+  belongs_to :condition
+  belongs_to :send_charge
+  belongs_to :prefecture
+  belongs_to :send_span
+
+    #空の投稿を保存できないようにする
+  validates :price, numericality: { only_integer: true, message: "is invalid. Input half-width characters." }
+  #'カテゴリー'、'商品状態'、'発送料の負担'、'発送元の地域'、'発送までの日数'の選択が「--」の時は保存できないようにする
+  validates :category_id, numericality: { other_than: 0, message: "can't be blank" }
+  validates :condition_id, numericality: { other_than: 0, message: "can't be blank" }
+  validates :send_charge_id, numericality: { other_than: 0, message: "can't be blank" }
+  validates :prefecture_id, numericality: { other_than: 0, message: "can't be blank" }
+  validates :send_span_id, numericality: { other_than: 0, message: "can't be blank" }
+
+  # 「寄付金」の金額範囲に関するバリデーション
+  validates :price, numericality: { greater_than_or_equal_to: 300, less_than_or_equal_to: 9999999, message: "is out of setting range" }
+end
```
こちらも商品出品機能の実装となりますので、一旦コメントアウトしておきましょう！

chobi555 replied:
こちらの記述は本実装とは関係のないものになるので、一度コメントアウトしておきましょう！

haramiyo commented on app/models/send_charge.rb:
```diff
@@ -0,0 +1,11 @@
+class send_charge < ActiveHash::Base
+  self.data = [
+    { id: 0, name: '--' },
+    { id: 1, name: '着払い（購入者負担）' },
+    { id: 2, name: '送料込み（出品者負担）' }
+  ]
+ 
+   include ActiveHash::Associations
+   has_many :products
+ 
```
こちらも商品出品機能の実装となりますので、一旦コメントアウトしておきましょう！

haramiyo commented on app/models/send_span.rb:
```diff
@@ -0,0 +1,12 @@
+class send_span < ActiveHash::Base
+  self.data = [
+    { id: 0, name: '--' },
+    { id: 1, name: '1~2日で発送' },
+    { id: 2, name: '2~3日で発送' },
+    { id: 3, name: '4~7日で発送' }
+  ]
+ 
+   include ActiveHash::Associations
+   has_many :products
```
こちらも商品出品機能の実装となりますので、一旦コメントアウトしておきましょう！

haramiyo commented on config/routes.rb:
```diff
@@ -1,6 +1,16 @@
   devise_for :users
-  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+  root to: 'products#index'
+  resources :users, only: [:edit, :update]
```
こちら必要なルーティングでしょうか？
もし不要であれば削除しておきましょう。

haramiyo commented on db/migrate/20210208143215_create_products.rb:
```diff
@@ -0,0 +1,18 @@
+  def change
+    create_table :products do |t|
+
+      t.timestamps
+      t.string :product_name, null: false
+      t.text :feature, null: false
+      t.integer :price, null: false
+      t.integer :category_id, null: false
+      t.integer :condition_id, null: false
+      t.integer :send_charge_id, null: false
+      t.integer :prefecture_id, null: false
+      t.integer :send_span_id, null: false
+      t.references :user, foreign_key: true
```
こちらも商品出品機能の実装となりますので、一旦コメントアウトしておきましょう！

haramiyo commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,83 @@
+      fill_in 'family_name_kana', with: ""
+      fill_in 'first_name_kana', with: ""
+      fill_in 'birth_profile', with: ""
+      expect{find('input[name="commit"]').click}.to change { User.count }.by(0)
+      expect(current_path).to eq "/users"
+    end
+  end
+end
+
+RSpec.describe 'ログイン', type: :system do
+  before do
+    @user = FactoryBot.create(:user)
+  end
+
+  context 'ログインができるとき' do
+    it '保存されているユーザーの情報と合致すればログインができる' do
+      visit root_path
+      expect(page).to have_content('ログイン')
+      visit new_user_session_path
+      fill_in 'Email', with: @user.email
+      fill_in 'Password', with: @user.password
+      find('input[name="commit"]').click
+      expect(current_path).to eq root_path
+      expect(find(".user_nav").find("span").hover).to have_content('ログアウト')
+      expect(page).to have_no_content('新規登録')
+      expect(page).to have_no_content('ログイン')
+    end
+  end
+
+  context 'ログインができないとき' do
+    it '保存されているユーザーの情報と合致しないとログインができない' do
+      visit root_path
+      expect(page).to have_content('ログイン')
+      visit new_user_session_path
+      fill_in 'Email', with: ""
+      fill_in 'Password', with: ""
+      find('input[name="commit"]').click
+      expect(current_path).to eq new_user_session_path
+    end
+  end
```
こちら結合テストとなっております。
最終課題ではモデルの単体テストが必須実装となっており、結合テストは任意実装となっております。

また、結合テストはspec/systemディレクトリ内に生成するファイルとなりますので、こちらの結合テストは一旦spec/systemディレクトリ内に移動していただき、modelの単体テスト用のファイルの作成をお願いいたします。

haramiyo reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、ご確認をお願いいたします！

また、こちらのプルリクエストですが、実際の画面や機能ができているか確認させていただきたいので、下記キャプチャ画像をご提出いただけますか？

・新規登録に成功する動画
・新規登録に失敗し、エラーが表示される動画
・ログインに成功する動画
・ログインに失敗し、エラーが表示される動画
・ログアウトに成功する動画
・テスト結果の画像

「Gyazo　使い方」とgoogleで検索し、ツールによって生成されたURLをこのPRに貼っていただければ幸いです。
その後、再度レビュー依頼フォームにてご提出をお願いいたします。

K-Hir-pk18 committed:
RSpec前

K-Hir-pk18 committed:
ユーザー管理機能（RSPEC）

##WHAT
・RSPECの単体テスト

##WHY
・ユーザー管理機能のテストのため

K-Hir-pk18 committed:
ユーザー管理（コメント対応）

WHAT
・各種コメントの対応

WHY
・可読性の向上

K-Hir-pk18 commented:
RSPECの結果です。
https://gyazo.com/0d5f1555ee89ecf83cb417647b2a994f

GYAZOのGIFはうまく撮れなかったので、、、


よろしくお願いいたします。

Katsuno0814 commented on spec/models/user_spec.rb:
```diff
@@ -1,84 +1,217 @@
+      it '正しい情報を入力すればユーザー新規登録ができてトップページに移動する' do
+        #visit root_path
+        #expect(page).to have_content('新規登録')
+        visit new_user_registration_path
```
> visit new_user_registration_path

こちらsystemのテストではないでしょうか？
こちらのファイルはモデルの単体ですとですので不要になります。
他記述でもsystemのテストを行われている箇所は削除いたしましょう！

K-Hir-pk18 replied:
katsuno0814様

対応しました。
path名を記述しないとRSPECが作動しないと思っていました。

path名なくて良いのですね！

Katsuno0814 reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いします！

K-Hir-pk18 committed:
ユーザー管理機能　コメント対応２

WHAT
RSPEC単体テストの修正

/spec/models/user_spec.rb
```
visit new_user_registration_path
```
箇所のコメントアウト

WHY
この記述があると、結合テストになるため

Katsuno0814 commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,217 @@
+    end
+  end
+
+  describe 'ログイン', type: :system do
+    before do
+      @user = FactoryBot.create(:user)
+    end
+
+    context 'ログインができるとき' do
+      it '保存されているユーザーの情報と合致すればログインができる' do
+        visit root_path
+        expect(page).to have_content('ログイン')
+        visit new_user_session_path
+        fill_in 'user[email]', with: @user.email
+        fill_in 'user[password]', with: @user.password
+        find('input[name="commit"]').click
+        expect(current_path).to eq root_path
+        expect(page).to have_content('ログアウト')
+        expect(page).to have_no_content('新規登録')
+        expect(page).to have_no_content('ログイン')
+      end
+    end
+  end
```
こちらの記述もお手数ですが削除いたしましょう！

K-Hir-pk18 replied:
Katsuno0814様

承知いたしました！

Katsuno0814 reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
コメントいたしましたので、お手数かと思いますがご確認お願いします。

K-Hir-pk18 committed:
ユーザー管理　コメント対応３

WHAT
・RSPECテスト”ログインができるとき”
のコメントアウト

WHY
・結合テストなので

K-Hir-pk18 commented:
よろしくお願いいたします！

chobi555 commented on config/routes.rb:
```diff
@@ -1,6 +1,8 @@
-  resources :products, only: :index
+  root to: "products#index"
+  resources :users
+  resources :products, only: [:index]
```
` resources :products, only: [:index]` 
本実装とは関係のないものになるので、一度コメントアウトしておきましょう！

chobi555 commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,217 @@
+require 'rails_helper'
```
こちらのファイル内に存在するコメントアウトは
必要ないかと思いますので、消しておきましょう！

K-Hir-pk18 replied:
最終行

```
describe 'ログイン', type: :system do
    before do
      @user = FactoryBot.create(:user)
    end
```

こちらの記述は不要だと思うのですが、いかがでしょうか。

chobi555 replied:
上記記述は必要でございますので残していただくよう、お願いいたします。

chobi555 reviewed (commented):
レビュー依頼ありがとうございます！
コメント致しました。
確認後修正をお願いいたします！

K-Hir-pk18 committed:
ユーザー管理機能　コメント対応４

WHAT
・models本機能に関係のないmodel
コメントアウトを実施

・routes.rb
products部のコメントアウト

・spec/modelsuser_spec.rb
コメントアウト部の削除

WHY
・可読性の向上

K-Hir-pk18 commented:
よろしくお願いいたします！

chobi555 commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,138 @@
+        @user.valid?
+        expect(@user.errors.full_messages).to include("Password is too short (minimum is 6 characters)")
+      end
+    
```
`パスワードが全角だと新規登録できない`をテスト項目に追記しましょう！

K-Hir-pk18 replied:
chobi555様

ご対応ありがとうございます。
ログイン部の記述は必要なんですね。
承知いたしました。

パスワード全角〜　のコード記述しました。
よろしくお願いいたします。

chobi555 reviewed (commented):
レビュー依頼ありがとうございます！
コメント致しました。
確認後修正をお願いいたします！

K-Hir-pk18 committed:
ユーザー管理機能　コメント対応５

WHAT
・RSPECテストの追加
　パスワードが全角だと登録できない

WHY
・基本機能の確認

K-Hir-pk18 commented:
よろしくお願いいたします！

Katsuno0814 reviewed (commented):
レビュー依頼並びに度々の修正ありがとうございます。
修正点確認いたしました。
LGTMとさせていただきます。実装お疲れ様でした。

マージ後、ローカルのmasterブランチに忘れずにプルしましょう。
その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
引き続き実装頑張っていきましょう。
